### ISSUE TITLE
Memory leak in MaxPool2d

### ISSUE THREAD

**Issue by m-bula** [pytorch/pytorch#32284]
## 🐛 Bug

It seems like there is very small memory leak during forward and backward propagation through network that can lead to memory overflow after many hours of training.

I stumbled upon this bug during A3C reinforcement training with multiple agents(processes) and reproduced it in a simple case in code underneath.

## To Reproduce

1.Run code below and observe memory consumption of process.

    import os
    os.environ["OMP_NUM_THREADS"] = "1"
    import torch
    import torch.multiprocessing as mp
    from setproctitle import setproctitle as ptitle
    import torch
    import torch.optim as optim
    import gc
    from torch import nn
    import torch.nn.functional as F
    from memory_profiler import profile

    class Network(torch.nn.Module):
        def __init__(self, num_inputs, num_outputs):
            super(Network, self).__init__()
            self.conv1 = nn.Conv2d(num_inputs, 32, 5, stride=1, padding=2)
            self.maxp1 = nn.MaxPool2d(2, 2)
            self.conv2 = nn.Conv2d(32, 64, 3, stride=1, padding=1)
            self.maxp2 = nn.MaxPool2d(2, 2)
            self.linear = nn.Linear(64*20*20, num_outputs)
            self.train()

        def forward(self, inputs):
            x = F.relu(self.maxp1(self.conv1(inputs)))
            x = F.relu(self.maxp2(self.conv2(x)))
            x = x.view(x.size(0), -1)
            return self.linear(x)

    def debug_memory():
        import resource
        print('maxrss = {}'.format(resource.getrusage(resource.RUSAGE_SELF).ru_maxrss))
        

    @profile
    def func():
        model = Network(1,6)
        optimizer = torch.optim.Adam(model.parameters(),lr=0.0001)
        mse_loss = nn.MSELoss()
        x = torch.randn([1, 1, 80, 80])
        y_ref = torch.randn([1, 6])
        while True:
            y = model(x)
            loss = mse_loss(y,y_ref)
            model.zero_grad()
            loss.backward(retain_graph=False)
            optimizer.step()
            del y, loss
            #debug_memory()

    if __name__ == '__main__':
        ptitle('MemoryLeakTest')
        func()

## Expected behavior
Memory consumption should have constant value. Instead, it slowly but steadily increases over time as shown below:
![memoryLeakTest-withenv-cropped](https://user-images.githubusercontent.com/59955198/72518980-c4e66d80-3856-11ea-82ac-b1bfba4d90f7.png)

I ran more tests and figured out that setting OMP_NUM_THREADS environment variable to 1 increases memory leakage. If i don't set it, memory increase is slower but still happens.
Here is a little comparison of memory used by process (data captured using `top`) with and without OMP_NUM_THREAD env variable set :

| OMP_NUM_THREADS=1        | After startup           | After 16 hours  |
| ------------- |:-------------:| -----:|
| set      | 216440 KB | 1.632 GB |
| not set      |  219404 KB      |   455616 KB |

When OMP_NUM_THREADS is not set memory increase isn't observable in short period of time ( less than 1h), but it will begin to show after few hours as shown in table.

## Environment

PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: Could not collect

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: No
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1060 6GB
Nvidia driver version: 430.26
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.17.4
[pip] torch==1.3.1
[pip] torchvision==0.4.2
[conda] _pytorch_select           0.1                       cpu_0  
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      243  
[conda] mkl-service               2.3.0            py36he904b0f_0  
[conda] mkl_fft                   1.0.15           py36ha843d7b_0  
[conda] mkl_random                1.1.0            py36hd6b4f25_0  
[conda] pytorch                   1.3.1           cpu_py36h62f834f_0  
[conda] torchvision               0.4.2                py36_cu101    pytorch

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen

**Comment by mrshenli** [comment]
Hi @m-bula does the memory consumption also increases if you drop the backward pass the optimizer step pass (and what if only drop the optimizer step pass)? That would help to narrow down the memory leak.

**Comment by m-bula** [comment]
Hi @mrshenli 
If I drop the optimizer step memory still keeps increasing:
![memoryLeakTest-withenv_no_optimzer_step_cropped](https://user-images.githubusercontent.com/59955198/72594916-ee5fd180-3908-11ea-93c8-22adfeb73b8d.png)

And if I put only forward pass in a loop the leak still happens as shown below:
![memoryLeakTest-withenv_onlyforward_cropped](https://user-images.githubusercontent.com/59955198/72595759-d5f0b680-390a-11ea-8856-2cbf473b4893.png)


So it seems like problem occurs mainly in forward pass, but I wouldn't be sure if it happens only there. I don't know how to check if it happens also in backward pass, because in order to do backward propagation I have to do forward propagation.

**Comment by rmcavoy** [comment]
I tried to par down the example code. The memory leak appears to come from MaxPool2d. I tried same example with relu+conv2d+linear with no maxpooling and there was no leak. I am on PyTorch 1.3.1, Ubuntu 18.04, Python3.7

```python3
import os
os.environ["OMP_NUM_THREADS"] = "1"
import torch
from torch import nn
from memory_profiler import profile
import resource

class Network(torch.nn.Module):
    def __init__(self):
        super(Network, self).__init__()
        self.maxp1 = nn.MaxPool2d(2, 2)

    def forward(self, x):
        return self.maxp1(x)

def debug_memory():
    print('maxrss = {}'.format(resource.getrusage(resource.RUSAGE_SELF).ru_maxrss))

@profile
def func():
    batch = 2
    channels = 1
    side_dim = 80
    model = Network()
    x = torch.randn([batch, channels, side_dim, side_dim])
    while True:
        y = model(x)
        debug_memory()
```

**Comment by mrshenli** [comment]
Thanks @m-bula  @rmcavoy for helping locate the memory leak.

Hey @bddppq @z-a-f @goldsborough, do you see any suspicious in [`max_pool2d`](https://github.com/pytorch/pytorch/blame/master/aten/src/ATen/native/Pooling.cpp#L120) ?

**Comment by m-bula** [comment]
@rmcavoy thank you for your help. I tried my code without maxpooling and indeed there is no memory leak, even after few hours.

**Comment by ezyang** [comment]
Bumping up priority as MaxPool2d is a very important operation

**Comment by albanD** [comment]
@mrshenli note that this is not linked to `max_pool2d`. In the for loop, replacing it with `torch._C._nn.max_pool2d_with_indices(x, 1)`, `torch._C._nn.log_sigmoid(x)`, `torch.threshold(x, 0, 0)` or `torch.relu(x)` all give the same behavior for me.

Given this, could this be a similar issue as https://github.com/pytorch/vision/issues/984#issuecomment-498603088 ?

**Comment by ezyang** [comment]
Probably not, since the reporter mentions:

> can lead to memory overflow after many hours of training

The malloc issue is not supposed to result in overflow.

If all of these functions are affected, we may be leaking a Python object. Maybe someone can get a heap dump of the process after we've leaked a pile of stuff and see what the leaked memory looks like.

**Comment by m-bula** [comment]
To be clear: by memory overflow I meant that program memory consumption eventually exceeds RAM resources, then tries to use SWAP memory and finally when all memory resources are gone system doesn't respond and needs to be restarted.

**Comment by rmcavoy** [comment]
@albanD Do you see a difference in the rate of leaking between MaxPool2d and ReLu? When I try ReLu, it either isn't leaking or leaks much more slowly than MaxPool2d. With ReLu, there is an initial step upwards but then it doesn't increase again after minutes of run time (millions of iterations). MaxPool2d leaks so quickly that I see it increasing every second or so.

**Comment by albanD** [comment]
I used a modified version of your code
```python
import os
import torch
torch.set_num_threads(1)
from torch import nn
from memory_profiler import profile
import resource
import gc


def debug_memory():
    val = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    # print('maxrss = {}'.format(val))
    return val

def fn(x):
    # choose one
    # torch._C._nn.max_pool2d_with_indices(x, 1)
    torch.relu(x)

@profile
def func():
    batch = 2
    channels = 1
    side_dim = 80
    x = torch.randn([batch, channels, side_dim, side_dim])
    old_val = debug_memory()
    for i in range(1000):
        fn(x)
        new_val = debug_memory()
        if new_val != old_val:
            print("Increased at {}: {}".format(i, new_val - old_val))
            old_val = new_val

func()
```

Both of them print between 3-10 increases in the `1000` iterations. It depends a lot from one run to the next. You don't see the same behavior?

**Comment by rmcavoy** [comment]
@albanD Mine currently prints only every 1000 iterations and I saw nothing with relu after an initial increase. Running 1000 iterations isn't really enough to diagnose this memory leak at all. This leak occurs over much longer time scales with even maxpool2d not leaking that fast. Maxpool2d leaks on the order of seconds which is tens of thousands of iterations of just maxpool(1,1)

**Comment by rmcavoy** [comment]
@albanD If you repeat the tests with the below code do you see real leaking behavior with all the functions you mentioned? My tests show only the two MaxPool2d have leaks (i.e. memory keeps increasing in the long term). The ReLu, threshold and log_sigmoid show memory increases in the first 1000 iterations (first two print statements) but are completely flat thereafter which is not a worrisome as the maxpool behavior.

```python
import os
os.environ["OMP_NUM_THREADS"] = "1"
import torch
import torch.nn.functional as F
from torch import nn
from memory_profiler import profile
import resource

class Network(torch.nn.Module):
    def __init__(self):
        super(Network, self).__init__()
        self.maxp1 = nn.MaxPool2d(1, 1)

    def forward(self, x):
        #return torch.relu(x)
        #return torch.threshold(x, 0, 0)
        #return torch._C._nn.log_sigmoid(x)
        return torch._C._nn.max_pool2d_with_indices(x, 1)
        #return self.maxp1(x)

def debug_memory():
    print('maxrss = {}'.format(resource.getrusage(resource.RUSAGE_SELF).ru_maxrss))

@profile
def func():
    batch = 2
    channels = 1
    side_dim = 80
    model = Network()
    x = torch.randn([batch, channels, side_dim, side_dim])
    for i in range(100000):
        y = model(x)
        if i % 1000 == 0:
            debug_memory()

if __name__ == '__main__':
    func()
```

**Comment by albanD** [comment]
I see the same thing with `log_sigmoid` for example. In my run, there was a memory increase in the last 10 prints.
But here again, I can't reproduce it with any of these on my centos machine. So this is very dependent on the machine.

**Comment by rmcavoy** [comment]
To confirm, are you referring to your run where you only do 1000 iterations or with the above script where it runs 100K iterations? The 1K (or even 2K) behavior is not a leak.

**Comment by albanD** [comment]
I am referring to the script you posted just above. So with `log_sigmoid` I still see increases in the last 30k of the 100k run.

**Comment by rmcavoy** [comment]
Interesting, is it continously increasing or only a relatively small amount late in the script? This seems quite insidious whatever it is.

**Comment by albanD** [comment]
It is similar amount compared to the beginning.
Have you tried using `jemalloc` to see what the behavior is on your machine?

**Comment by ezyang** [comment]
This is high priority to try the jemalloc fix to try and see if it is a memory allocator error.

**Comment by mattip** [comment]
I could confirm the leak on that exact version (v1.3.1 with numpy 1.17.4, no cuda, python3.6). By only changing the pytorch version to v1.4.0, the leak no longer occurs. Should I bisect it further to figure out exactly where the change occurred?

**Comment by mattip** [comment]
While v1.4.0 does not leak, current master (d83509e) does.

**Comment by mattip** [comment]
I think this is due to a quirk in `omp_in_parallel()` and the use of nested `parallel for`s  in [`max_pool2d_with_indices_out_frame`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/DilatedMaxPool2d.cpp#L105). According to [this discussion](https://software.intel.com/en-us/forums/intel-moderncode-for-parallel-architectures/topic/282426), `omp_in_parallel()` will only be true if we are already in a parallel context with **more than one thread**. But in the case at hand, we only have one thread, so `omp_in_parallel()` returns false, and we end up inside a double `parallel for` which somehow messes up OpenMP's memory management. I am not sure why that should leak memory, but if I modify the `#pragma omp parallel` in [`ParallelOpenMP.h`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelOpenMP.h#L29) to check `num_threads > 1` then the leak goes away.

I understand the nested parallel for works on other `at::parallelfor` implementations, so we cannot remove it, so we must fix the OpenMP parallel for pragma.

**Comment by albanD** [comment]
cc @VitalyFedyunin do you have an idea why this could happen?

**Comment by ezyang** [comment]
@mattip  that seems like a pretty reasonable diagnosis. Does this mean we have to manually keep track of nested OMP ourselves?

**Comment by mattip** [comment]
PR gh-36479 proposes a solution without any manual tracking.

**Comment by tranthanhsongitrepo** [comment]
I'm encountering this problem in version 1.9.0

**Comment by rgommers** [comment]
@tranthanhsongitrepo please open a new issue with enough detail that it can be investigated.

**Comment by tranthanhsongitrepo** [comment]
> @tranthanhsongitrepo please open a new issue with enough detail that it can be investigated.

I've found the problem elsewhere, the memory usage was rising when I call max pool but I've failed to notice that I still have some hidden reference to it down the line
